apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,901 @@
+                        status: StatusCode,
+                    ) -> Result<ControlFlow<router::Response, router::Request>, BoxError>
+                    {
+                        // This is a metric and will not appear in the logs
+                        tracing::info!(
+                            monotonic_counter.apollo_authentication_failure_count = 1u64,
+                            kind = %AUTHENTICATION_KIND
+                        );
+                        let response = router::Response::error_builder()
+                            .error(
+                                graphql::Error::builder()
+                                    .message(msg)
+                                    .extension_code("AUTH_ERROR")
+                                    .build(),
+                            )
+                            .status_code(status)
+                            .context(context)
+                            .build()?;
+                        Ok(ControlFlow::Break(response))
+                    }
+
+                    // The http_request is stored in a `Router::Request` context.
+                    // We are going to check the headers for the presence of the configured header
+                    let jwt_value_result =
+                        match request.router_request.headers().get(&my_config.header_name) {
+                            Some(value) => value.to_str(),
+                            None => {
+                                return failure_message(
+                                    request.context,
+                                    format!("Missing {} header", &my_config.header_name),
+                                    StatusCode::UNAUTHORIZED,
+                                );
+                            }
+                        };
+
+                    // If we find the header, but can't convert it to a string, let the client know
+                    let jwt_value_untrimmed = match jwt_value_result {
+                        Ok(value) => value,
+                        Err(_not_a_string_error) => {
+                            return failure_message(
+                                request.context,
+                                "configured header is not convertible to a string".to_string(),
+                                StatusCode::BAD_REQUEST,
+                            );
+                        }
+                    };
+
+                    // Let's trim out leading and trailing whitespace to be accommodating
+                    let jwt_value = jwt_value_untrimmed.trim();
+
+                    // Make sure the format of our message matches our expectations
+                    // Technically, the spec is case sensitive, but let's accept
+                    // case variations
+                    if !jwt_value
+                        .to_uppercase()
+                        .as_str()
+                        .starts_with(&my_config.header_prefix.to_uppercase())
```

Comment:
Should we strip the header_prefix ?

Reply:
I'm not sure what you intend. This is a case insensitive comparison of the header prefix with the expected configured value. How would stripping the `header_value_prefix` help with that?

Reply:
It was more to avoid errors like the user has written 1 or 2 spaces at the end of the prefix in the yaml configuration and doesn't understand why it's not working. But maybe we should not think about that and don't touch a this prefix at all.